### PR TITLE
Summary variables calculated only when called

### LINKED ISSUE #4058: Calculate summary variables only when they are called
As discussed in #3995 

- create a `SummaryVariable` object that saves `solution.last_state` and uses it to calculate the esoh variables only when they are called (the same way as `Solution` calculates the variables only when they are called)
- allow users to pass in functions that calculate summary variables based on other variables

### PR DESCRIPTION
**Pull request by pipliggins**
# Description

Rather than calculating summary variables and storing them in every Solution class, a 'SummaryVariables' class is created which stores the first + last state of the solution, plus any relevant esoh parameters, so that the summary variables are only calculated and stored when they are called (similar to accessing standard variables). If an esoh variable is requested, they are all calculated and stored at once to reduce the number of calls to the esoh solver.

rather than viewing the list of summary variable names as
```
sol.summary_variables.keys()
```
you can now use
```
sol.summary_variables.all_variables
```
to view the names without calculating the variables.

Similarly, when plotting rather than getting the number of cycles as 
```
sol.summary_variables['Cycle number']
```
they are accessed as 
```
sol.summary_variables.cycle_number
```

Fixes #4058 

## Type of change

- [x] Optimization (back-end change that speeds up the code)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: src/pybamm/solvers/summary_variable.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        self.user_inputs = user_inputs or {}
+        self.esoh_solver = esoh_solver
+        self._variables = {}  # Store computed variables
+        self.cycle_number = None
+
+        model = solution.all_models[0]
+        self._possible_variables = model.summary_variables  # minus esoh variables
+        self._esoh_variables = []  # Store eSOH variable names
+
+        # Flag if eSOH calculations are needed
```
Comment (rtimms) [inline src/pybamm/solvers/summary_variable.py:44]: related: https://github.com/pybamm-team/PyBaMM/issues/4619

File: src/pybamm/solvers/summary_variable.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            self.first_state = solution.first_state
+            self.last_state = solution.last_state
+            self.cycles = None
+
+    def _initialize_for_cycles(self, cycle_summary_variables: list[SummaryVariables]):
+        """Initialize attributes for when multiple cycles are provided."""
+        self.first_state = None
+        self.last_state = None
+        self.cycles = cycle_summary_variables
+        self.cycle_number = np.arange(1, len(self.cycles) + 1)
```
Comment (martinjrobins) [inline src/pybamm/solvers/summary_variable.py:66]: I can see that at least `cycle_number` is used outside this class, can you please document the attributes that are meant to be public in the docstring?

**Review by martinjrobins — CHANGES_REQUESTED** [review]
sorry for the delay in the review! This looks great, can you please add to the CHANGELOG under breaking changes, since this changes the api for the summary variables?

**Review by martinjrobins — APPROVED** [review]
looks good, thanks @pipliggins

**Pull request merged**